agent/service_manager.go

```diff
@@ -379,39 +379,76 @@ func mergeServiceConfig(defaults *structs.ServiceConfigResponse, service *struct
 	if ns.Proxy.MeshGateway.Mode == structs.MeshGatewayModeDefault {
 		ns.Proxy.MeshGateway.Mode = defaults.MeshGateway.Mode
 	}
+	if !ns.Proxy.TransparentProxy {
+		ns.Proxy.TransparentProxy = defaults.TransparentProxy
+	}
+
+	// seenUpstreams stores the upstreams seen from the local registration so that we can also add synthetic entries.
+	// for upstream configuration that was defined via service-defaults.UpstreamConfigs. In TransparentProxy mode
+	// ns.Proxy.Upstreams will likely be empty because users do not need to define upstreams explicitly.
+	// So to store upstream-specific flags from central config, we add entries to ns.Proxy.Upstream with thosee values.
+	remoteUpstreams := make(map[structs.ServiceID]structs.Upstream)
+
+	for _, us := range defaults.UpstreamIDConfigs {
+		parsed, err := structs.ParseUpstreamConfig(us.Config)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse upstream config map for %s: %v", us.Upstream.String(), err)
+		}
 
-	// Merge upstream defaults if there were any returned
+		// Delete the mesh gateway key since this is the only place it is read from an opaque map.
+		// Later reads use Proxy.MeshGateway.
+		delete(us.Config, "mesh_gateway")
```

Comment:
Would you also need to delete variations of this (camel, UPPER, etc) or is that safe to ignore since this is coming straight back from the RPC endpoint?

Reply:
Safe to ignore for the reason you just mentioned. That RPC endpoint and `UpstreamConfig.MergeInto()` are the only places that populate mesh gateway config in the opaque map.

Added a comment about that.